consul/src/main/java/com/linecorp/armeria/common/consul/ConsulClient.java

```diff
@@ -0,0 +1,277 @@
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+import com.linecorp.armeria.internal.consul.AgentServiceClient;
+import com.linecorp.armeria.internal.consul.CatalogClient;
+import com.linecorp.armeria.internal.consul.ConsulParameters;
+import com.linecorp.armeria.internal.consul.HealthClient;
+
+/**
+ * The Consul Client for accessing to consul agent API server.
+ */
+public class ConsulClient {
+    /**
+     * Default Consul API SCHEME.
+     */
+    public static final String DEFAULT_SCHEME = "http";
+
+    /**
+     * Default Consul HTTP API HOST.
+     */
+    public static final String DEFAULT_HTTP_HOST = "localhost";
+
+    /**
+     * Default Consul HTTP API PORT.
+     */
+    public static final int DEFAULT_HTTP_PORT = 8500;
+
+    /**
+     * Default Consul API version.
+     */
+    public static final String DEFAULT_API_VERSION = "v1";
+
+    private static final Logger logger = LoggerFactory.getLogger(HealthClient.class);
+
+    @Nullable
+    private final String token;
+    private final WebClient webClient;
+
+    protected ConsulClient(String uri) {
+        this(uri, null);
+    }
+
+    protected ConsulClient(String uri, @Nullable String token) {
+        webClient = WebClient.of(uri);
+        this.token = token;
+    }
+
+    /**
+     * Registers a service to Consul Agent without service ID.
+     *
+     * @param serviceName service name
+     * @param endpoint servicing endpoint
+     * @param check a check for health checking
+     * @return CompletableFuture with registered service ID(auto-generated)
+     */
+    public CompletableFuture<String> register(String serviceName, Endpoint endpoint, Check check) {
+        return AgentServiceClient.of(this)
+                                 .register(serviceName, endpoint.host(), endpoint.port(), check,
+                                           ConsulParameters.BLANK);
+    }
+
+    /**
+     * Registers a service to Consul Agent with service ID.
+     *
+     * @param serviceId a service ID that identifying a service
+     * @param serviceName a service name to register
+     * @param endpoint an endpoint of service to register
+     * @param check a check for the service
+     * @return CompletableFuture with registered service ID
+     */
+    public CompletableFuture<String> register(String serviceId, String serviceName, Endpoint endpoint,
+                                              Check check) {
+        return AgentServiceClient.of(this)
+                                 .register(serviceId, serviceName, endpoint.host(), endpoint.port(), check,
+                                           ConsulParameters.BLANK);
+    }
+
+    /**
+     * De-registers a service to Consul Agent.
+     *
+     * @param serviceId a service ID that identifying a service
+     */
+    public CompletableFuture<Void> deregister(String serviceId) {
+        return AgentServiceClient.of(this).deregister(serviceId);
```

Comment:
Could you explain more about `AgentServiceClient`, `CatalogClient` and `HealthClient` what they are for, please?

Reply:
Of course @minwoox ! They are mapped to the endpoints of the Consul API. `AgentServiceClient` is responsible for [`/agent/service`](https://www.consul.io/api/agent/service.html). `CatalogClient` is responsible for [`/catalog`](https://www.consul.io/api/catalog.html). `HealthClient` is responsible for [`/health`](https://www.consul.io/api/health.html). I also wrote these to javadoc comments.

Reply:
If `AgentServiceClient`, `CatalogClient` and `HealthClient` are mapped to Consul REST API, 
How about merging them into `ConsulClient` except the entity models? So we can simplify our logic a lot.

Reply:
I think the opposite is true. This API has a similar domain model, but apparently different, which was very confusing. So I thought it was better to separate each one.
The reason is that there is some code that is not being used now. But it will be useful in the future.
Let me know if you have a good idea. I will be happy to follow Armeria's policy.